Add regions of interest highlighting

pletcher opened the pull request:
This feature adds bidirectional highlighting (to-from the folio image) when regions of interest are available for text and/or scholia.

I think this is almost ready. I still need to map clicking on the ROI in the folio image to selecting the scholion in the sidebar, and there's some weird behavior when clicking scholia now that we've added additional data: it seems like clicking a scholion in the folio will also highlight other regions of interest?

I'll need to investigate where they're coming from.


![scholia-highlights](https://github.com/scaife-viewer/frontend/assets/2015049/062e4659-09bd-4b80-83c9-a88bb9cc1bc1)

pletcher commented:
I think this is basically ready to go, but I'd love to confirm that the interactions and highlight colors work as expected.

CI is failing on `yarn install`:

```
error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Error: The process '/usr/local/bin/yarn' failed with exit code 1
    at ExecState._setResult (/home/runner/work/_actions/bahmutov/npm-install/v1.4.3/dist/index.js:936:25)
    at ExecState.CheckComplete (/home/runner/work/_actions/bahmutov/npm-install/v1.4.3/dist/index.js:919:18)
    at ChildProcess.<anonymous> (/home/runner/work/_actions/bahmutov/npm-install/v1.4.3/dist/index.js:819:27)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
Error: The process '/usr/local/bin/yarn' failed with exit code 1
```

pletcher commented on packages/widget-scholia/src/Scholion.vue:
```diff
@@ -0,0 +1,77 @@
+      // NB: After the highlightedScholion is set for the first time,
+      // it becomes an Observer --- that's why we check for null/undefined
+      // of highlightedScholion.idx as well.
+      if (highlightedScholion !== null && highlightedScholion.idx != null) {
```
I'm not sure about checks like this one -- I'm not familiar with the idioms for dealing with Vuex's Observer objects, but this feels a little hacky to me.

pletcher commented on packages/widget-commentary/src/CommentaryWidget2.vue:
I think a lot of files, including this one, snuck in during a rebase.

pletcher commented:
Okay, I believe the latest changes have fixed the bugs identified in the demo earlier this week.

I've added a couple of comments -- it looks like there's some git cleanup from rebasing. Definitely my bad, but not sure if it's necessary to fix manually since we can just squash these down?

jacobwegner commented:
@pletcher: Thanks for the update!

I do see some of the "rebase noise" you're talking about, and I think I remember you were using a newer Yarn version that we were working through some issues with earlier in the year.

I'll pull this branch and take a look tomorrow.

jacobwegner commented:
@pletcher I've been pulled over to other things for Beyond Translation and scaife.perseus.org, but hope to circle back here after our meeting today.

I apologize for the continued delay!

pletcher commented:
No problem!

pletcher commented:
Sorry that took a bit, but I fixed the minor bugs that showed up in the demo this morning and cleaned up the selection state on the highlighter. Thanks in advance for the review!

jacobwegner commented on src/main.js:
```diff
@@ -57,6 +58,7 @@ Vue.use(SkeletonPlugin, {
     ...audioIconMap,
     ...neIconMap,
     ...deIconMap,
+    ...tocIconMap,
```
Thanks for this fix!

jacobwegner commented on packages/reader-image-mode/src/ImageViewer.vue:
```diff
@@ -88,6 +143,83 @@
         }
       },
+      async clearHighlights() {
+        await this.$store.dispatch(`${MODULE_NS}/${HIGHLIGHT_TRANSCRIPTION}`, {
```
I changed this from a mutation to an action for consistency with other usage of Vuex in the project.

jacobwegner commented on packages/widget-scholia/src/Scholion.vue:
```diff
@@ -0,0 +1,88 @@
+
+    .highlight {
+      background: var(
+        --sv-widget-reader-token-selected-entity-shadow-color,
```
This is fine for now, but we may want to circle back and either:

1) Make image reader mode specific css variables
2) Generalize the `--sv-widget-reader-token-` variable names (since we're using them other places)

jacobwegner commented on packages/widget-reader/src/ReaderTextPart.vue:
```diff
@@ -32,16 +36,25 @@
+
         this.$store.dispatch(`${MODULE_NS}/${SELECT_LINE}`, {
-          ref: `${this.urn.version}${this.textPart.ref}`,
+          ref: this.textPart.ref,
```
This change is ok for now, but if we were working with the `SELECT_LINE` action in a context where say, we were showing two different versions, I'd prefer a "fully-qualified" reference (e.g., the text part URN).

pletcher replied:
Ah, right -- this was a bit shortsighted on my part. I was trying to avoid having to parse the URN in multiple places, but your approach makes more sense. I can definitely fix this in a subsequent pass.

jacobwegner reviewed (approved):
@pletcher Again, with great apologies on the long lag time between when you've started all of this work and when I've finished it...

I pushed a couple of small commits to ensure consistency with the rest of the codebase, but am otherwise eager to bring these changes in!

I made a few comments in the PR review that might be things to keep in mind for future PRs.

Great work!
